File: python/semx/survival.py
import numpy as np
import pandas as pd
from typing import Union, Sequence, Mapping, List, Optional
from .model import SemFit

def predict_survival(
    fit: SemFit, 
    data: Union[Mapping[str, Sequence[float]], pd.DataFrame], 
    times: Sequence[float],
    outcome: str
) -> pd.DataFrame:
    """Predict survival probabilities S(t) for a given outcome.
    
    Parameters
    ----------
    fit : SemFit
        Fitted model.
    data : pd.DataFrame
        New data for prediction.
    times : Sequence[float]
        Time points to predict survival at.
    outcome : str
        Name of the survival outcome variable.
        
    Returns
    -------
    pd.DataFrame
        DataFrame where columns are time points and rows are observations.
        Values are S(t).
    """
    # Find variable definition
    var_def = next((v for v in fit.model_ir.variables if v.name == outcome), None)
    if not var_def:
        raise ValueError(f"Outcome '{outcome}' not found in model.")
        
    family = var_def.family
    
    # Predict linear predictor eta
    # fit.predict returns predictions for all endogenous variables
    # We need to ensure we get the linear predictor for 'outcome'
    preds = fit.predict(data)
    if outcome not in preds.columns:
        raise ValueError(f"Could not predict linear predictor for '{outcome}'")
        
    eta = preds[outcome].values
    
    # Get dispersion parameter (shape)
    # Try "psi_{outcome}_{outcome}" first
    dispersion_id = f"psi_{outcome}_{outcome}"
    dispersion = fit.parameter_estimates.get(dispersion_id)
    
    if dispersion is None:
        if family == "exponential":
            dispersion = 1.0
        else:
            # Try fuzzy match
            candidates = [k for k in fit.parameter_estimates if outcome in k and ("psi" in k or "shape" in k)]
            if len(candidates) == 1:
                dispersion = fit.parameter_estimates[candidates[0]]
            else:
                # Default to 1.0 if not found? Or raise?
                # For Weibull, shape=1 is Exponential.
                # Let's warn and use 1.0? No, raise.
                raise ValueError(f"Could not identify dispersion parameter for '{outcome}' (family: {family}). Candidates: {candidates}")
    
    k = dispersion
    
    results = {}
    for t in times:
        if t < 0:
            results[t] = np.nan
            continue
            
        # Weibull AFT: S(t) = exp( - (t * exp(-eta))^k )
        # Log-logistic: S(t) = 1 / (1 + (t * exp(-eta))^k)
        # Lognormal: S(t) = 1 - Phi( (log(t) - eta) / k ) ? (k is sigma)
        
        if family == "weibull" or family == "exponential":
            # z = (t * exp(-eta))^k
            # log(z) = k * (log(t) - eta)
            log_t = np.log(t) if t > 0 else -np.inf
            log_z = k * (log_t - eta)
            z = np.exp(log_z)
            S_t = np.exp(-z)
        elif family == "loglogistic":
            # S(t) = 1 / (1 + (t * exp(-eta))^k)
            log_t = np.log(t) if t > 0 else -np.inf
            log_z = k * (log_t - eta)
            z = np.exp(log_z)
            S_t = 1.0 / (1.0 + z)
        elif family == "lognormal":
            from scipy.stats import norm
            # T ~ Lognormal(mu=eta, sigma=k)
            # S(t) = 1 - CDF((log(t) - eta)/k)
            if t <= 0:
                S_t = 1.0
            else:
                z = (np.log(t) - eta) / k
                S_t = 1.0 - norm.cdf(z)
        else:
            raise ValueError(f"Unsupported family for survival prediction: {family}")
            
        results[t] = S_t
        
    return pd.DataFrame(results)

def predict_cif(
    fit: SemFit,
    data: Union[Mapping[str, Sequence[float]], pd.DataFrame],
    times: Sequence[float],
    event_outcome: str,
    competing_outcomes: Sequence[str]
) -> pd.DataFrame:
    """Predict Cumulative Incidence Function (CIF) for a specific event.
    
    CIF_j(t) = integral_0^t S_overall(u) * h_j(u) du
    
    Parameters
    ----------
    fit : SemFit
        Fitted model.
    data : pd.DataFrame
        New data.
    times : Sequence[float]
        Time points.
    event_outcome : str
        The outcome variable for the event of interest.
    competing_outcomes : Sequence[str]
        List of outcome variables for competing events.
        
    Returns
    -------
    pd.DataFrame
        CIF values.
    """
    # We need predictions for all outcomes
    all_outcomes = [event_outcome] + list(competing_outcomes)
    
    # Get parameters for all outcomes
    params = {} # outcome -> {eta: vector, k: scalar, family: str}
    
    preds = fit.predict(data)
    n_obs = len(preds)
    
    for out in all_outcomes:
        var_def = next((v for v in fit.model_ir.variables if v.name == out), None)
        if not var_def:
            raise ValueError(f"Outcome '{out}' not found")
        
        eta = preds[out].values
        
        dispersion_id = f"psi_{out}_{out}"
        dispersion = fit.parameter_estimates.get(dispersion_id)
        if dispersion is None:
             if var_def.family == "exponential":
                 dispersion = 1.0
             else:
                 candidates = [k for k in fit.parameter_estimates if out in k and ("psi" in k or "shape" in k)]
                 if len(candidates) == 1:
                     dispersion = fit.parameter_estimates[candidates[0]]
                 else:
                     raise ValueError(f"Dispersion for {out} not found")
                     
        params[out] = {"eta": eta, "k": dispersion, "family": var_def.family}

    # Numerical integration
    # We compute S_overall(u) and h_event(u) on a grid
    # Grid: 0 to max(times)
    max_time = max(times)
    # Use fine grid? Or just the requested times if they are dense?
    # Let's use a fixed grid for integration and interpolate?
    # Or just integrate step-wise between requested times (assuming sorted).
    
    sorted_times = sorted([t for t in times if t >= 0])
    if not sorted_times:
        return pd.DataFrame()
        
    # Add 0 if not present
    integration_points = sorted(list(set([0.0] + sorted_times)))
    
    cif_values = np.zeros((n_obs, len(integration_points)))
    
    # CIF(0) = 0
    
    for i in range(1, len(integration_points)):
        t_prev = integration_points[i-1]
        t_curr = integration_points[i]
        dt = t_curr - t_prev
        t_mid = (t_prev + t_curr) / 2.0
        
        # Compute S_overall(t_mid) and h_event(t_mid)
        
        # S_overall = product S_j
        S_overall = np.ones(n_obs)
        for out in all_outcomes:
            p = params[out]
            # Compute S_j(t_mid)
            # Reuse logic from predict_survival (inline for speed)
            k = p["k"]
            eta = p["eta"]
            fam = p["family"]
            
            if fam in ["weibull", "exponential"]:
                log_z = k * (np.log(t_mid) - eta)
                z = np.exp(log_z)
                S_j = np.exp(-z)
            elif fam == "loglogistic":
                log_z = k * (np.log(t_mid) - eta)
                z = np.exp(log_z)
                S_j = 1.0 / (1.0 + z)
            elif fam == "lognormal":
                from scipy.stats import norm
                z_score = (np.log(t_mid) - eta) / k
                S_j = 1.0 - norm.cdf(z_score)
            else:
                S_j = np.ones(n_obs) # Should not happen
            
            S_overall *= S_j
            
        # h_event(t_mid)
        p_evt = params[event_outcome]
        k = p_evt["k"]
        eta = p_evt["eta"]
        fam = p_evt["family"]
        
        if fam in ["weibull", "exponential"]:
            # h(t) = k/t * (t*exp(-eta))^k
            #      = k/t * z
            log_z = k * (np.log(t_mid) - eta)
            z = np.exp(log_z)
            h_evt = (k / t_mid) * z
        elif fam == "loglogistic":
            # h(t) = (k/t * z) / (1+z)
            log_z = k * (np.log(t_mid) - eta)
            z = np.exp(log_z)
            h_evt = (k / t_mid) * z / (1.0 + z)
        elif fam == "lognormal":
            from scipy.stats import norm
            # h(t) = pdf / S
            z_score = (np.log(t_mid) - eta) / k
            pdf = norm.pdf(z_score) / (t_mid * k)
            S_val = 1.0 - norm.cdf(z_score)
            h_evt = pdf / S_val
        else:
            h_evt = np.zeros(n_obs)
            
        # CIF increment
        d_cif = S_overall * h_evt * dt
        cif_values[:, i] = cif_values[:, i-1] + d_cif
        
    # Map back to requested times
    # We computed on integration_points which includes requested times
    # We just need to extract columns corresponding to requested times
    
    # Create mapping
    time_to_idx = {t: i for i, t in enumerate(integration_points)}
    
    final_results = {}
    for t in times:
        if t < 0:
            final_results[t] = np.nan
        else:
            idx = time_to_idx[t]
            final_results[t] = cif_values[:, idx]
            
    return pd.DataFrame(final_results)

File: Rpkg/semx/R/survival.R
#' Predict survival probabilities
#'
#' @param fit A semx_fit object.
#' @param newdata A data.frame.
#' @param times Numeric vector of times.
#' @param outcome Name of the outcome variable.
#' @return A data.frame of survival probabilities.
#' @export
semx_predict_survival <- function(fit, newdata, times, outcome) {
  model <- fit$model
  params <- fit$optimization_result$parameters
  names(params) <- model$ir$parameter_ids()
  
  # Find family
  var_def <- model$variables[[outcome]]
  if (is.null(var_def)) stop(sprintf("Outcome '%s' not found", outcome))
  family <- var_def$family
  
  # Predict eta
  preds <- predict(fit, newdata)
  if (is.null(preds[[outcome]])) stop(sprintf("Could not predict for '%s'", outcome))
  eta <- preds[[outcome]]
  
  # Find dispersion
  dispersion_id <- sprintf("psi_%s_%s", outcome, outcome)
  k <- params[[dispersion_id]]
  
  if (is.null(k)) {
    if (family == "exponential") {
      k <- 1.0
    } else {
      # Fuzzy search
      candidates <- grep(paste0("psi.*", outcome, "|shape.*", outcome), names(params), value = TRUE)
      if (length(candidates) == 1) {
        k <- params[[candidates]]
      } else {
        stop(sprintf("Dispersion parameter for '%s' not found", outcome))
      }
    }
  }
  
  results <- matrix(NA, nrow = length(eta), ncol = length(times))
  colnames(results) <- as.character(times)
  
  for (j in seq_along(times)) {
    t <- times[j]
    if (t < 0) next
    
    if (family %in% c("weibull", "exponential")) {
      log_z <- k * (log(t) - eta)
      z <- exp(log_z)
      S_t <- exp(-z)
    } else if (family == "loglogistic") {
      log_z <- k * (log(t) - eta)
      z <- exp(log_z)
      S_t <- 1.0 / (1.0 + z)
    } else if (family == "lognormal") {
      z_score <- (log(t) - eta) / k
      S_t <- 1.0 - pnorm(z_score)
    } else {
      stop(sprintf("Unsupported family: %s", family))
    }
    results[, j] <- S_t
  }
  
  as.data.frame(results)
}

#' Predict Cumulative Incidence Function
#'
#' @param fit A semx_fit object.
#' @param newdata A data.frame.
#' @param times Numeric vector of times.
#' @param event_outcome Name of the event outcome.
#' @param competing_outcomes Vector of competing outcome names.
#' @return A data.frame of CIF values.
#' @export
semx_predict_cif <- function(fit, newdata, times, event_outcome, competing_outcomes) {
  all_outcomes <- c(event_outcome, competing_outcomes)
  preds <- predict(fit, newdata)
  n_obs <- nrow(newdata)
  
  params <- fit$optimization_result$parameters
  names(params) <- fit$model$ir$parameter_ids()
  
  outcome_params <- list()
  for (out in all_outcomes) {
    var_def <- fit$model$variables[[out]]
    family <- var_def$family
    eta <- preds[[out]]
    
    dispersion_id <- sprintf("psi_%s_%s", out, out)
    k <- params[[dispersion_id]]
    if (is.null(k)) {
      if (family == "exponential") k <- 1.0
      else {
         candidates <- grep(paste0("psi.*", out, "|shape.*", out), names(params), value = TRUE)
         if (length(candidates) == 1) k <- params[[candidates]]
         else stop(sprintf("Dispersion for %s not found", out))
      }
    }
    outcome_params[[out]] <- list(eta = eta, k = k, family = family)
  }
  
  sorted_times <- sort(unique(c(0, times[times >= 0])))
  cif_values <- matrix(0, nrow = n_obs, ncol = length(sorted_times))
  
  for (i in 2:length(sorted_times)) {
    t_mid <- (sorted_times[i-1] + sorted_times[i]) / 2
    dt <- sorted_times[i] - sorted_times[i-1]
    
    S_overall <- rep(1, n_obs)
    for (out in all_outcomes) {
      p <- outcome_params[[out]]
      k <- p$k; eta <- p$eta; fam <- p$family
      
      if (fam %in% c("weibull", "exponential")) {
        z <- exp(k * (log(t_mid) - eta))
        S_j <- exp(-z)
      } else if (fam == "loglogistic") {
        z <- exp(k * (log(t_mid) - eta))
        S_j <- 1 / (1 + z)
      } else if (fam == "lognormal") {
        S_j <- 1 - pnorm((log(t_mid) - eta) / k)
      }
      S_overall <- S_overall * S_j
    }
    
    p_evt <- outcome_params[[event_outcome]]
    k <- p_evt$k; eta <- p_evt$eta; fam <- p_evt$family
    
    if (fam %in% c("weibull", "exponential")) {
      z <- exp(k * (log(t_mid) - eta))
      h_evt <- (k / t_mid) * z
    } else if (fam == "loglogistic") {
      z <- exp(k * (log(t_mid) - eta))
      h_evt <- (k / t_mid) * z / (1 + z)
    } else if (fam == "lognormal") {
      z_score <- (log(t_mid) - eta) / k
      h_evt <- dnorm(z_score) / (t_mid * k * (1 - pnorm(z_score)))
    }
    
    d_cif <- S_overall * h_evt * dt
    cif_values[, i] <- cif_values[, i-1] + d_cif
  }
  
  final_results <- matrix(NA, nrow = n_obs, ncol = length(times))
  colnames(final_results) <- as.character(times)
  
  for (j in seq_along(times)) {
    idx <- match(times[j], sorted_times)
    final_results[, j] <- cif_values[, idx]
  }
  
  as.data.frame(final_results)
}
